File: jar/src/main/java/jline/api/pfqn/nc/Pfqn_ncoi.java
/**
 * @file Order-independent (OI) balanced-fairness normalizing constant.
 *
 * Computes the exact normalizing constant G(N) of a closed product-form
 * queueing network comprising an arbitrary number of order-independent (OI) /
 * pass-and-swap stations with empty swap graph and a single aggregated
 * infinite-server (delay) node. The OI stations are analyzed by the
 * balanced-fairness recursion of Bonald and Proutiere (2003), combined with
 * the multichain convolution over stations on the count lattice.
 *
 * Port of matlab/src/api/pfqn/pfqn_ncoi.m.
 *
 * @since LINE 3.0
 */
package jline.api.pfqn.nc;

import jline.io.Ret;
import org.apache.commons.math3.special.Gamma;

import java.util.ArrayList;
import java.util.List;
import java.util.function.ToDoubleFunction;

public final class Pfqn_ncoi {
    private Pfqn_ncoi() {}

    /**
     * Normalizing constant of a closed OI + single-delay product-form network.
     *
     * <p>The balance function of each OI station is the balanced-fairness
     * recursion Phi(0) = 1, Phi(n) = (1/mu(n)) sum_{r: n_r&gt;0} Phi(n - e_r),
     * and G(N) is the convolution of those balance functions with the
     * multinomial delay factor F_Z(n) = prod_r Z_r^{n_r}/n_r!.
     *
     * <p>This is a MACROSTATE routine: everything is tabulated over the count
     * lattice 0 &lt;= n &lt;= N, never over orderings, which is legitimate
     * because an OI rate is permutation-invariant so Phi closes on the count
     * vector. With a non-empty swap graph that closure fails and
     * {@link Pfqn_pas_nc} must be used instead. Cost: O(M R L) for the balance
     * functions and O(M prod_r (N_r+1)(N_r+2)/2) for the convolutions, with
     * L = prod_r (N_r+1).
     *
     * @param Z  (R) think-time demand of the aggregated delay node.
     * @param N  (R) closed population, finite.
     * @param mu list of OI station rate handles; each maps a per-class count
     *           vector n (length R) to the total service rate at that station.
     *           A state with non-positive rate is unreachable and gets a zero
     *           balance value. May be empty/null to model a pure delay network.
     * @return G(N) and log G(N).
     */
    public static Ret.pfqnOiNc pfqn_ncoi(double[] Z, int[] N, List<ToDoubleFunction<int[]>> mu) {
        return pfqn_ncoi(Z, N, mu, null);
    }

    // General per-OI-station class visit ratios: visits[m] is the 1xR visit
    // vector of OI station m, entering the v-weighted balanced-fairness balance
    // Phi^v(n)=(1/mu(n)) sum_r v_r Phi^v(n-e_r). null -> unit visits.
    public static Ret.pfqnOiNc pfqn_ncoi(double[] Z, int[] N, List<ToDoubleFunction<int[]>> mu, double[][] visits) {
        int R = N.length;
        List<ToDoubleFunction<int[]>> muList = (mu == null) ? new ArrayList<ToDoubleFunction<int[]>>() : mu;

        if (Z == null || Z.length == 0) {
            Z = new double[R];
        }
        if (Z.length != R) {
            throw new RuntimeException("pfqn_ncoi: Z and N must have the same number of classes.");
        }
        for (int r = 0; r < R; r++) {
            if (N[r] < 0) {
                throw new RuntimeException("pfqn_ncoi requires finite (closed) populations.");
            }
        }
        if (R == 0) {
            return new Ret.pfqnOiNc(1.0, 0.0, new double[]{1.0});
        }

        // Count lattice 0 <= n <= N, flattened column-major; lin(v) = sum(v*strides)
        // is linear, so lin(x+y) = lin(x) + lin(y), which the convolution exploits.
        int[] dims = new int[R];
        int[] strides = new int[R];
        int ngrid = 1;
        for (int r = 0; r < R; r++) {
            dims[r] = N[r] + 1;
            strides[r] = ngrid;
            ngrid *= dims[r];
        }

        int[][] counts = new int[ngrid][R];
        int total = 0;
        for (int r = 0; r < R; r++) {
            total += N[r];
        }
        // Lattice points bucketed by total population: the balanced-fairness
        // recursion must be swept in population-increasing order.
        List<List<Integer>> byPop = new ArrayList<List<Integer>>();
        for (int s = 0; s <= total; s++) {
            byPop.add(new ArrayList<Integer>());
        }
        for (int k = 0; k < ngrid; k++) {
            int rest = k;
            int sum = 0;
            for (int r = 0; r < R; r++) {
                counts[k][r] = rest % dims[r];
                rest /= dims[r];
                sum += counts[k][r];
            }
            byPop.get(sum).add(k);
        }

        // Delay balance function: the multinomial factor F_Z(n). A class with
        // population but no delay demand makes the state infeasible.
        double[] g = new double[ngrid];
        for (int k = 0; k < ngrid; k++) {
            double logf = 0.0;
            boolean feas = true;
            for (int r = 0; r < R; r++) {
                if (counts[k][r] > 0) {
                    if (Z[r] <= 0) {
                        feas = false;
                        break;
                    }
                    logf += counts[k][r] * Math.log(Z[r]) - Gamma.logGamma(counts[k][r] + 1.0);
                }
            }
            if (feas) {
                g[k] = Math.exp(logf);
            }
        }

        // Convolve in one OI station at a time.
        int[] rem = new int[R];
        int[] y = new int[R];
        for (int m = 0; m < muList.size(); m++) {
            double[] vism = (visits != null && m < visits.length) ? visits[m] : null;
            double[] phi = balance(counts, byPop, strides, muList.get(m), R, ngrid, vism);
            double[] gnext = new double[ngrid];
            for (int kx = 0; kx < ngrid; kx++) {
                double px = phi[kx];
                if (px == 0.0) {
                    continue;
                }
                int base = 0;
                for (int r = 0; r < R; r++) {
                    rem[r] = N[r] - counts[kx][r];
                    y[r] = 0;
                    base += counts[kx][r] * strides[r];
                }
                // Odometer over the sub-box 0 <= y <= rem.
                while (true) {
                    int ylin = 0;
                    for (int r = 0; r < R; r++) {
                        ylin += y[r] * strides[r];
                    }
                    gnext[base + ylin] += px * g[ylin];
                    int d = 0;
                    while (d < R && y[d] == rem[d]) {
                        y[d] = 0;
                        d++;
                    }
                    if (d == R) {
                        break;
                    }
                    y[d]++;
                }
            }
            g = gnext;
        }

        // g now holds G(n) for EVERY n on the lattice, not just n = N.
        double G = g[ngrid - 1];
        double lG = (G > 0) ? Math.log(G) : Double.NEGATIVE_INFINITY;
        return new Ret.pfqnOiNc(G, lG, g);
    }

    /**
     * Balanced-fairness recursion Phi(n) = (1/mu(n)) sum_{r: n_r&gt;0} Phi(n-e_r)
     * over the count lattice, swept in population-increasing order.
     */
    private static double[] balance(int[][] counts, List<List<Integer>> byPop, int[] strides,
                                    ToDoubleFunction<int[]> murate, int R, int ngrid, double[] vis) {
        double[] phi = new double[ngrid];
        for (int s = 0; s < byPop.size(); s++) {
            List<Integer> bucket = byPop.get(s);
            for (int b = 0; b < bucket.size(); b++) {
                int k = bucket.get(b);
                if (s == 0) {
                    phi[k] = 1.0;
                    continue;
                }
                double mun = murate.applyAsDouble(counts[k]);
                if (!(mun > 0)) {
                    continue; // unreachable station state: zero balance value
                }
                double acc = 0.0;
                for (int r = 0; r < R; r++) {
                    if (counts[k][r] > 0) {
                        double vr = (vis == null) ? 1.0 : vis[r];
                        acc += vr * phi[k - strides[r]];
                    }
                }
                phi[k] = acc / mun;
            }
        }
        return phi;
    }
}

File: matlab/src/api/pfqn/pfqn_ncoi.m
function [G, lG, Gtab] = pfqn_ncoi(Z, N, mu, visits, options)
% [G, LG, GTAB] = PFQN_NCOI(Z, N, MU, OPTIONS)
%
% Normalizing constant for a closed product-form queueing network that
% comprises a single aggregated infinite-server (delay) node and an
% arbitrary number of order-independent (OI) / pass-and-swap stations with
% empty swap graph.
%
% The OI stations are analyzed by the balanced-fairness recursion of
% Bonald & Proutiere (2003), "Insensitive bandwidth sharing in data
% networks", combined with the multichain convolution over stations. For a
% single OI station with rank rate mu(supp(n)) the balance function is
%   Phi(0) = 1,  Phi(n) = (1/mu(n)) * sum_{r: n_r>0} Phi(n - e_r),
% and G(N) is obtained by convolving the per-station balance functions with
% the multinomial delay factor F_Z(n) = prod_r Z_r^{n_r} / n_r!,
%   g_0(n) = F_Z(n),   g_m(n) = sum_{0<=x<=n} Phi_m(x) g_{m-1}(n-x),
%   G(N)   = g_M(N).
%
% This is a MACROSTATE routine: both the balance functions and the
% convolution are tabulated over the count lattice 0 <= n <= N, never over
% orderings. That is legitimate exactly because an OI rate is permutation-
% invariant, so Phi(n) -- itself the sum of the ordered-prefix weights
% prod_p 1/mu(c_1..c_p) over all orderings c of the multiset n -- closes on
% the count vector. With a non-empty swap graph that closure fails and the
% microstate routine PFQN_PAS_NC must be used instead.
%
% COST. With L = prod_r (N_r+1) the balance functions cost O(M R L) and the
% convolutions O(M sum_{n<=N} prod_r (n_r+1)) = O(M prod_r (N_r+1)(N_r+2)/2),
% i.e. the order of a load-dependent Buzen convolution: polynomial in the
% population for a fixed number of classes.
%
% Parameters:
%   Z  - (1 x R) think-time demand vector of the aggregated delay node.
%        Z(r) = 1/sigma_r for a delay with per-class rate sigma_r.
%   N  - (1 x R) closed population vector, finite.
%   mu - cell array {1 x M} of function handles, one per OI station. Each
%        mu{m}(n) returns the total service rate of station m given the
%        per-class occupancy (count) vector n (1 x R). For an order-
%        independent station this rate depends only on the support of n
%        (which classes are present), i.e. mu{m}(n) = sum of the capacities
%        of the servers compatible with the classes present in n. A station
%        state whose rate is non-positive is unreachable and is assigned a
%        zero balance value. May be empty to model a pure delay network.
%   options - solver options (optional, currently unused; accepted for
%             signature parity with the other pfqn_* routines).
%
% Returns:
%   G    - Normalizing constant G(N).
%   lG   - log(G(N)).
%   Gtab - (prod_r(N_r+1) x 1) the WHOLE lattice of normalizing constants,
%          Gtab(1 + sum(n .* strides)) = G(n) for every 0 <= n <= N, with
%          strides = [1, cumprod(N(1:end-1)+1)]. The convolution produces this
%          table anyway, so a caller that needs G at more than one population
%          (throughputs, queue lengths, a fold of further stations) must take
%          this output and index it, NOT re-call the routine per population --
%          the latter costs a needless factor prod_r(N_r+1).
%
% Example (IS + two OI stations, R classes):
%   oirate  = @(n) sum(mu1(any(compat1(:, find(n>0)) ~= 0, 2)));
%   oirate2 = @(n) sum(mu2(any(compat2(:, find(n>0)) ~= 0, 2)));
%   G = pfqn_ncoi(1./sigma, N, {oirate, oirate2});
%
% See also PFQN_PAS_NC, PFQN_OI_FNC, PFQN_OI_INSVC, PFQN_MVAOI.
%
% Copyright (c) 2012-2026, Imperial College London
% All rights reserved.

if nargin < 5
    options = struct(); %#ok<NASGU>
end
if nargin < 4
    visits = [];
end
if nargin < 3 || isempty(mu)
    mu = {};
end
if ~iscell(mu)
    mu = {mu};
end

R = numel(N);
if isempty(Z)
    Z = zeros(1, R);
end
if numel(Z) ~= R
    line_error(mfilename, 'Z and N must have the same number of classes.');
end
if any(~isfinite(N))
    line_error(mfilename, 'pfqn_ncoi requires finite (closed) populations.');
end

N = round(N(:)');
Z = Z(:)';
M = numel(mu);

% Per-OI-station class visit ratios. Default unit visits reproduce the plain
% rank-rate balance; general visits enter as a per-class weight on the
% balanced-fairness recurrence, Phi^v(n)=(1/mu(n)) sum_r v_r Phi^v(n-e_r), the
% v-weighted balance whose class-r geometric factor prod_r v_r^{n_r} carries
% the OI-station visit ratio (see _kb/06-solver-catalog.md, NC OI analyzer).
if nargin < 4 || isempty(visits)
    visits = repmat({ones(1, R)}, 1, M);
elseif ~iscell(visits)
    vmat = visits; visits = cell(1, M);
    for mm = 1:M, visits{mm} = vmat(mm, :); end
end

% Count lattice 0 <= n <= N, flattened column-major; lin(v) = 1 + sum(v.*strides)
% is affine, so lin(x+y) = lin(x) + lin(y) - 1, which the convolution exploits.
dims = N + 1;
ngrid = prod(dims);
strides = [1, cumprod(dims(1:end-1))];
% mixed radix rather than ind2sub, which rejects a single-class (scalar) dims
counts = zeros(ngrid, R);
lin0 = (0:ngrid-1)';
for r = 1:R
    counts(:, r) = mod(floor(lin0 / strides(r)), dims(r));
end
[~, ord] = sort(sum(counts, 2)); % population-increasing sweep order

% Delay balance function: the multinomial factor F_Z(n). A class with
% population but no delay demand makes the state infeasible (weight zero).
g = zeros(ngrid, 1);
for k = 1:ngrid
    v = counts(k, :);
    logf = 0;
    feas = true;
    for r = 1:R
        if v(r) > 0
            if Z(r) <= 0
                feas = false;
                break
            end
            logf = logf + v(r) * log(Z(r)) - gammaln(v(r) + 1);
        end
    end
    if feas
        g(k) = exp(logf);
    end
end

% Convolve in one OI station at a time.
for m = 1:M
    Phim = oi_nc_balance(counts, ord, strides, mu{m}, R, ngrid, visits{m});
    gnext = zeros(ngrid, 1);
    for kx = 1:ngrid
        px = Phim(kx);
        if px == 0
            continue
        end
        x = counts(kx, :);
        rem = N - x;
        % Linear indices of the sub-box 0 <= y <= rem, built by mixed radix.
        ylin = 1;
        for d = 1:R
            ylin = bsxfun(@plus, ylin(:), strides(d) * (0:rem(d)));
            ylin = ylin(:);
        end
        base = 1 + sum(x .* strides);
        gnext(base + ylin - 1) = gnext(base + ylin - 1) + px * g(ylin);
    end
    g = gnext;
end

% g now holds G(n) for EVERY n on the lattice, not just n = N.
Gtab = g;
G = g(ngrid);
lG = log(G);
end

function Phi = oi_nc_balance(counts, ord, strides, murate, R, ngrid, vis)
% v-weighted balanced-fairness recursion
% Phi(n) = (1/mu(n)) sum_{r: n_r>0} v_r Phi(n-e_r), over the count lattice,
% swept in population-increasing order. vis is the 1xR class visit vector.
if nargin < 7 || isempty(vis), vis = ones(1, R); end
Phi = zeros(ngrid, 1);
for t = 1:ngrid
    k = ord(t);
    n = counts(k, :);
    if all(n == 0)
        Phi(k) = 1;
        continue
    end
    mun = murate(n);
    if ~(mun > 0)
        continue % unreachable station state: zero balance value
    end
    acc = 0;
    for r = 1:R
        if n(r) > 0
            acc = acc + vis(r) * Phi(k - strides(r));
        end
    end
    Phi(k) = acc / mun;
end
end
